compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -876,6 +876,52 @@ impl<'tcx> PredicateVisitor<'tcx> for HasEscapingVarsVisitor {
     }
 }
 
+crate struct CountBoundVars {
+    crate outer_index: ty::DebruijnIndex,
+    crate bound_tys: FxHashSet<ty::BoundTy>,
+    crate bound_regions: FxHashSet<ty::BoundRegion>,
+    crate bound_consts: FxHashSet<ty::BoundVar>,
+}
+
+impl<'tcx> TypeVisitor<'tcx> for CountBoundVars {
+    fn visit_binder<T: TypeFoldable<'tcx>>(&mut self, t: &Binder<T>) -> bool {
+        self.outer_index.shift_in(1);
+        let result = t.super_visit_with(self);
+        self.outer_index.shift_out(1);
+        result
+    }
+
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        match t.kind {
+            ty::Bound(debruijn, ty) if debruijn >= self.outer_index => {
```

Comment:
Why do we consider for `for<T> for<U> Struct<T, U>` that the inner binder (`for<U>`) binds `T` and `U` while for
`for<'a> for<'b> Struct<'a, 'b>` the inner binder only binds `b` afaict?

In case I am misunderstanding something here: why do we use `debruijn >= self.outer_index` for types and `debruijn == self.outer_index` for regions?

Reply:
This was a typo. I did fix it in a later commit on my other branch. They aren't here, because I only wanted to test the minimal set of changes for perf. But yeah, good catch!